plugins/chat/app/services/chat/auto_join_channels.rb

```diff
@@ -0,0 +1,106 @@
+
+      sql = <<~SQL
+        WITH chat_users AS ( -- users that are allowed to join chat
+          SELECT u.id
+            FROM users u
+            JOIN user_options uo ON uo.user_id = u.id AND uo.chat_enabled = TRUE
+           WHERE u.id > 0
+             AND u.active = TRUE
+             AND u.staged = FALSE
+             AND (u.suspended_till IS NULL OR u.suspended_till <= :now)
+             AND (u.silenced_till IS NULL OR u.silenced_till <= :now)
+             AND NOT EXISTS (SELECT 1 FROM anonymous_users au WHERE au.user_id = u.id)
+             AND u.last_seen_at > :last_seen_at
+             #{user_sql}
+             #{everyone_allowed ? "" : "AND EXISTS (SELECT 1 FROM group_users gu WHERE gu.user_id = u.id AND gu.group_id IN (:group_ids))"}
+           ORDER BY u.last_seen_at DESC
+           LIMIT :max_users
+        ), valid_chat_channels AS ( -- auto joinable chat channels
+          SELECT cc.id, cc.chatable_id
+            FROM chat_channels cc
+           WHERE cc.auto_join_users = TRUE
+             AND cc.chatable_type = 'Category'
+             AND cc.deleted_at IS NULL
+             AND cc.user_count < :max_users
+             #{channel_sql}
+        ), public AS ( -- public chat channels
+          SELECT cu.id AS user_id, cc.id AS chat_channel_id
+            FROM valid_chat_channels cc
+            CROSS JOIN chat_users cu
+            JOIN categories c ON c.id = cc.chatable_id AND c.read_restricted = FALSE #{category_sql}
+        ), private AS ( -- private chat channels
+          SELECT DISTINCT gu.user_id, cc.id AS chat_channel_id
+            FROM valid_chat_channels cc
+            JOIN categories c ON c.id = cc.chatable_id AND c.read_restricted = TRUE #{category_sql}
+            JOIN category_groups cg ON cg.category_id = c.id AND cg.permission_type IN (:group_permissions)
+            JOIN group_users gu ON gu.group_id = cg.group_id AND gu.user_id IN (SELECT id FROM chat_users)
+        )
+        INSERT INTO user_chat_channel_memberships (user_id, chat_channel_id, following, join_mode, created_at, updated_at)
+        SELECT p.user_id, p.chat_channel_id, TRUE, :automatic, :now, :now
+          FROM (
+            SELECT * FROM public 
+            UNION ALL 
+            SELECT * FROM private
+          ) p
+          LEFT JOIN user_chat_channel_memberships uccm ON uccm.user_id = p.user_id AND uccm.chat_channel_id = p.chat_channel_id
+         WHERE uccm.user_id IS NULL
+        RETURNING chat_channel_id, user_id
+      SQL
+
+      channel_to_user = Hash.new { |h, k| h[k] = [] }
+      args = { now:, last_seen_at:, group_ids:, max_users:, group_permissions:, automatic: }
+
+      DB
+        .query_array(sql, args)
+        .each { |channel_id, user_id| channel_to_user[channel_id] << user_id }
+
+      ::Chat::Channel
+        .where(id: channel_to_user.keys)
+        .find_each do |channel|
+          users = User.where(id: channel_to_user[channel.id])
```

Comment:
This can end up loading a maximum of 10K ActiveRecord objects at once which will bloat memory. Looking at `::Chat::Publisher.publish_new_channel(channel, users)`, we don't actually need to create ActiveRecord objects here and can just pass the user_ids to `::Chat::Publisher.publish_new_channel` with some tweaks.

Reply:
You sure it's loading the records? This is just an AR object that is being passed around and only being used in `publish_new_channel` in a `where` clause. Rails is clever enough to only extract the ids.

I guess I could make that explicit.